dask/dataframe/core.py

```diff
@@ -1130,21 +1166,33 @@ def _bind_operator_method(cls, name, op):
         def meth(self, other, level=None, fill_value=None, axis=0):
             if not level is None:
                 raise NotImplementedError('level must be None')
-            return map_partitions(op, self.column_info, self, other,
+            return map_partitions(op, self._pd, self, other,
                                   axis=axis, fill_value=fill_value)
         meth.__doc__ = op.__doc__
         bind_method(cls, name, meth)
 
     def apply(self, func, convert_dtype=True, name=no_default, args=(), **kwds):
         """ Parallel version of pandas.Series.apply """
+
         if name is no_default:
-            name = self.name
+            msg = ("name is not specified, inferred from partial data. "
+                   "Please provide name if the result is unecpected.\n"
+                   "  Before: s.apply(func)\n"
+                   "  After:  s.apply(func, name=['x', 'y']) for dataframe result\n"
+                   "  or:     s.apply(func, name='x')        for series result")
+            warnings.warn(msg)
+
+            name = _emulate(pd.Series.apply, self.head(), func,
+                            convert_dtype=convert_dtype,
+                            args=args, **kwds)
```

Comment:
Hrm, ok, this behavior surprised me:

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'x': [1, 2, 3, 4], 'y': [10, 20, 30, 40]})

In [3]: df
Out[3]: 
   x   y
0  1  10
1  2  20
2  3  30
3  4  40

In [4]: def return_df(x):
   ...:             # will create new DataFrame which columns is ['sum', 'mean']   ...:             return pd.Series([x.sum(), x.mean()], index=['sum', 'mean'])   ...: 

In [5]: df.apply(return_df, axis=1)
Out[5]: 
    sum  mean
0  11.0   5.5
1  22.0  11.0
2  33.0  16.5
3  44.0  22.0

In [6]: df.iloc[0:0].apply(return_df, axis=1)
Out[6]: 
Empty DataFrame
Columns: [x, y]
Index: []
```

I expected Pandas to produce a dataframe with the same information regardless of the length of the length of the input.  I guess that Pandas needs some data to determine the output type though.